ci: Use rust toolchain actions

MichaReiser opened the pull request:
Use the `actions-rs/toolchain` action to install Rust, `actions-rs/cargo` to run cargo commands, and `action-rs/clippy-check` to run clippy.

* Using `actinos-rs/toolchain` ensures that our workflows work with [act](https://github.com/nektos/act), allowing to test locally
* Using `actions-rs/cargo`: It gives nice warnings and errors in the GitHub UI
* Using `actions/clippy-check`: Posts lint annotations

## Test Plan

Sample output after I added clippy errors

https://github.com/charliermarsh/ruff/actions/runs/4404224020/jobs/7713601234

charliermarsh commented:
Seems reasonable to me. We intentionally moved away from `actions-rs/toolchain` at Boshen's recommendation based on profiling work from Rome. I'm curious if your benchmarking shows that it's much slower (or not).

Boshen commented:
The benefit is that you don't have to sync your CI Rust versions with rust-toolchain.toml.

MichaReiser commented:
:wave: @Boshen Thanks for chiming in. 

> The benefit is that you don't have to sync your CI Rust versions with rust-toolchain.toml.

This is no longer an issue because the action now supports the [toolchain file](https://github.com/actions-rs/toolchain#the-toolchain-file)

> This Action supports [toolchain files](https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file), so it is not necessary to use toolchain input anymore. ....

charliermarsh commented:
@Boshen heard me even without the tag

stinodego commented:
For what it's worth, we moved away from `actions-rs/toolchain` as it is [unmaintained](https://github.com/actions-rs/toolchain/issues/216). We moved to [`dtolnay/rust-toolchain`](https://github.com/dtolnay/rust-toolchain), which seems to work fine.

MichaReiser commented:
> For what it's worth, we moved away from `actions-rs/toolchain` as it is [unmaintained](https://github.com/actions-rs/toolchain/issues/216). We moved to [`dtolnay/rust-toolchain`](https://github.com/dtolnay/rust-toolchain), which seems to work fine.

Hmm, that's a good point. I believe [`dtolnay/rust-toolchain`](https://github.com/dtolnay/rust-toolchain) doesn't respect the version specified in the `toolchain.toml`. But we could give [setup-rust-toolchain](https://github.com/actions-rust-lang/setup-rust-toolchain) a try.
